### PR TITLE
Issue #506: Implemented a lower-footprint row annotation factory

### PR DESCRIPTION
**Pull request by kaspersorensen**
Attempt to "fix"/improve #513

So far I did not find any obvious greedy "mistakes" or something like that in the Pattern Finder. But I did find that it depends a lot on the RowAnnotationFactory. And here we might improve a bit on the memory footprint.

Note: I am still testing this solution, do not merge yet. Will post test results on this PR's comments.

### CODE REVIEW

**Comment by kaspersorensen** [comment]
__results of using this PR__

 * Scenario 1:
  * 1.8 M records with 35 fields
  * 4 pattern finders used on the first 4 fields
  * Execution time: 1,172 min
  * Allocated memory after run and after GC: 1,467,904 gb

**Comment by kaspersorensen** [comment]
__results from latest DC release (4.0.8)__

 * Scenario 1:
  * (same input file, job etc.)
  * Execution time: 1,065 min
  * Allocated memory after run and after GC: 1,559,552 kb

**Comment by kaspersorensen** [comment]
New test run, 1,5 M record file with 5 fields (the old file had a bunch of inconsistencies which made it a strage test file)

After the new PR:
image

**Comment by kaspersorensen** [comment]
With the previous implementation we're above the 750 MB tick:

image

**Comment by kaspersorensen** [comment]
More importantly, the new implementation let's you tweak it. In conf.xml you can now play around with these values to increase/decrease sampling in memory:

```
			<row-annotation-storage>
				<in-memory max-rows-threshold="500" max-sets-threshold="500"/>
			</row-annotation-storage>
```

**Comment by kaspersorensen** [comment]
Case in point, here's the memory profile when running with just 10 sets and a threshold of 10 records:

image

**Comment by stefanjanssen** [comment]
About the last two evaluations: If I understand the threshold thing right, it seems as if the rows are in memory even when there is no job running?

File: engine/core/src/main/java/org/datacleaner/storage/AbstractRowAnnotationFactory2.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+package org.datacleaner.storage;
+
+import java.util.List;
+
+import org.datacleaner.api.InputRow;
+
+/**
+ * Abstract implementation of {@link RowAnnotationSampleContainer}
+ */
+public abstract class AbstractRowAnnotationFactory2 implements RowAnnotationFactory, RowAnnotationSampleContainer,
```
Comment (stefanjanssen) [inline engine/core/src/main/java/org/datacleaner/storage/AbstractRowAnnotationFactory2.java]: Maybe a more descriptive name than Factory2 would be nice? The javadoc doesn't give any insight in why this is number two either.

File: engine/core/src/main/java/org/datacleaner/storage/InMemoryRowAnnotationFactory2.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        }
+    }
+
+    private Collection<InputRow> getInputRowCollection(int defaultSize, RowAnnotation annotation) {
+        List<InputRow> rowCollection = _storage.get(annotation);
+        if (rowCollection == null) {
+            if (_storage.size() >= _maxSampleSets) {
+                return null;
+            }
+            rowCollection = new ArrayList<InputRow>(defaultSize);
```
Comment (stefanjanssen) [inline engine/core/src/main/java/org/datacleaner/storage/InMemoryRowAnnotationFactory2.java]: Strictly speaking this is not properly concurrent: two threads can modify the same ArrayList at the same time. ArrayList is not synchronized.

**Comment by stefanjanssen** [comment]
Other than the remarks above, the changes look good to me. If you aim at reducing the memory footprint even further I would probably take a look at the InputRow class, as this class is (by nature) referencing a lot of those.

File: engine/core/src/main/java/org/datacleaner/storage/InMemoryRowAnnotationFactory2.java (hunk 2 above)
Comment (stefanjanssen) [inline engine/core/src/main/java/org/datacleaner/storage/InMemoryRowAnnotationFactory2.java]: Also, ConcurrentHashMap has a setting for the maximum number of concurrent modifications, which you can tweak (or just put to one) to optimize memory usage as it is doing a concurrency/memory trade off.

**Comment by kaspersorensen** [comment]
Good remarks. Thanks.

**Pull request merged**